File: frontend/src/engine/nfaEngine.ts
import { NFAMachine, NFATransition, SimulationResult, SimulationStepTrace } from '../types/automata';

export const isEpsilon = (symbol: string): boolean => {
  return !symbol || symbol === 'ε' || symbol === 'eps' || symbol === 'EPS' || symbol === 'E' || symbol === 'λ';
};

/**
 * Computes the ε-closure for a set of NFA states
 */
export function computeEpsilonClosure(states: string[], transitions: NFATransition[]): string[] {
  const closure = new Set<string>(states);
  const queue = [...states];

  while (queue.length > 0) {
    const current = queue.shift()!;
    const epsTransitions = transitions.filter(t => t.from === current && isEpsilon(t.symbol));

    for (const t of epsTransitions) {
      if (!closure.has(t.to)) {
        closure.add(t.to);
        queue.push(t.to);
      }
    }
  }

  return Array.from(closure).sort();
}

/**
 * Simulates an NFA on an input string
 */
export function simulateNFA(machine: NFAMachine, inputString: string): SimulationResult {
  const traces: SimulationStepTrace[] = [];

  if (!machine.startState) {
    return {
      accepted: false,
      finalStatus: 'ERROR',
      traces: [{
        stepIndex: 0,
        activeStates: [],
        remainingInput: inputString,
        inputIndex: 0,
        actionSummary: 'No start state defined for NFA',
        status: 'ERROR',
      }],
      totalSteps: 1,
      message: 'No start state defined for NFA.',
    };
  }

  // Initial step with ε-closure
  let currentStates = computeEpsilonClosure([machine.startState], machine.transitions);
  const stateNames = (ids: string[]) => ids.map(id => machine.states.find(s => s.id === id)?.label || id).join(', ');

  traces.push({
    stepIndex: 0,
    activeStates: [...currentStates],
    currentSymbol: inputString.length > 0 ? inputString[0] : undefined,
    inputIndex: 0,
    remainingInput: inputString,
    actionSummary: `Initial active states (after ε-closure of ${machine.startState}): {${stateNames(currentStates)}}`,
    status: 'RUNNING',
  });

  for (let i = 0; i < inputString.length; i++) {
    const symbol = inputString[i];
    const nextStatesSet = new Set<string>();
    const usedTransitionIds: string[] = [];

    for (const st of currentStates) {
      const directMatches = machine.transitions.filter(t => t.from === st && t.symbol === symbol);
      for (const t of directMatches) {
        nextStatesSet.add(t.to);
        usedTransitionIds.push(t.id);
      }
    }

    if (nextStatesSet.size === 0) {
      // Branch dead-end
      traces.push({
        stepIndex: i + 1,
        activeStates: [],
        currentSymbol: symbol,
        inputIndex: i,
        remainingInput: inputString.slice(i),
        actionSummary: `No transitions on '${symbol}' from active states {${stateNames(currentStates)}} (All branches halted)`,
        status: 'REJECTED',
      });

      return {
        accepted: false,
        finalStatus: 'REJECTED',
        traces,
        totalSteps: traces.length,
        message: `NFA rejected: No branches survived reading symbol '${symbol}'.`,
      };
    }

    // Apply ε-closure to the resulting states
    const directArray = Array.from(nextStatesSet);
    currentStates = computeEpsilonClosure(directArray, machine.transitions);
    
    const nextSymbol = i + 1 < inputString.length ? inputString[i + 1] : undefined;
    const remaining = inputString.slice(i + 1);

    traces.push({
      stepIndex: i + 1,
      activeStates: [...currentStates],
      currentSymbol: nextSymbol,
      inputIndex: i + 1,
      remainingInput: remaining,
      activeTransitionId: usedTransitionIds[0],
      actionSummary: `Read '${symbol}' → Reached {${stateNames(directArray)}} | ε-closure: {${stateNames(currentStates)}}`,
      status: 'RUNNING',
    });
  }

  // Check acceptance
  const isAccept = currentStates.some(st => machine.acceptStates.includes(st));
  const finalStatus = isAccept ? 'ACCEPTED' : 'REJECTED';

  if (traces.length > 0) {
    const lastTrace = traces[traces.length - 1];
    lastTrace.status = finalStatus;
    const acceptingStates = currentStates.filter(st => machine.acceptStates.includes(st));
    if (isAccept) {
      lastTrace.actionSummary += ` | Accepted via accepting state(s): {${stateNames(acceptingStates)}}`;
    } else {
      lastTrace.actionSummary += ` | Rejected: none of {${stateNames(currentStates)}} are accept states`;
    }
  }

  return {
    accepted: isAccept,
    finalStatus,
    traces,
    totalSteps: traces.length,
    message: isAccept
      ? `Input string "${inputString}" ACCEPTED by NFA.`
      : `Input string "${inputString}" REJECTED by NFA.`,
  };
}

File: backend/engine/nfa_engine.py
from typing import List, Set
try:
    from models.schemas import NFAMachine, SimulationResult, SimulationStepTrace
except ImportError:
    from ..models.schemas import NFAMachine, SimulationResult, SimulationStepTrace

def get_epsilon_closure(states: Set[str], machine: NFAMachine) -> Set[str]:
    closure = set(states)
    queue = list(states)
    while queue:
        curr = queue.pop(0)
        for t in machine.transitions:
            if t.from_state == curr and (t.symbol == 'ε' or t.symbol == '' or t.symbol == 'eps'):
                if t.to_state not in closure:
                    closure.add(t.to_state)
                    queue.append(t.to_state)
    return closure

def simulate_nfa(machine: NFAMachine, input_str: str) -> SimulationResult:
    traces: List[SimulationStepTrace] = []
    state_lookup = {s.id: s.label for s in machine.states}

    initial_closure = get_epsilon_closure({machine.startState}, machine)
    current_states = initial_closure
    chars = list(input_str)

    has_accept = any(s in machine.acceptStates for s in current_states)
    labels = [state_lookup.get(s, s) for s in sorted(current_states)]

    traces.append(
        SimulationStepTrace(
            stepIndex=0,
            activeStates=sorted(list(current_states)),
            remainingInput=input_str,
            readSymbol=None,
            activeTransitionId=None,
            actionSummary=f"Initial state closure: {{{', '.join(labels)}}}",
            status="RUNNING" if len(chars) > 0 else ("ACCEPTED" if has_accept else "REJECTED"),
        )
    )

    if len(chars) == 0:
        return SimulationResult(
            accepted=has_accept,
            finalStatus="ACCEPTED" if has_accept else "REJECTED",
            totalSteps=1,
            traces=traces,
            message=f"Empty string ε {'accepted' if has_accept else 'rejected'}.",
        )

    for i, sym in enumerate(chars):
        next_states: Set[str] = set()
        for curr in current_states:
            for t in machine.transitions:
                if t.from_state == curr and t.symbol == sym:
                    next_states.add(t.to_state)

        closure = get_epsilon_closure(next_states, machine)
        current_states = closure
        remaining = "".join(chars[i+1:])
        is_last = (i == len(chars) - 1)
        has_acc = any(s in machine.acceptStates for s in current_states)
        lbls = [state_lookup.get(s, s) for s in sorted(current_states)]

        if len(current_states) == 0:
            traces.append(
                SimulationStepTrace(
                    stepIndex=len(traces),
                    activeStates=[],
                    remainingInput=remaining,
                    readSymbol=sym,
                    activeTransitionId=None,
                    actionSummary=f"No valid branches for symbol '{sym}'. All paths terminated (∅).",
                    status="REJECTED",
                )
            )
            return SimulationResult(
                accepted=False,
                finalStatus="REJECTED",
                totalSteps=len(traces),
                traces=traces,
                message=f"Rejected: All active branches halted on '{sym}'.",
            )

        traces.append(
            SimulationStepTrace(
                stepIndex=len(traces),
                activeStates=sorted(list(current_states)),
                remainingInput=remaining,
                readSymbol=sym,
                activeTransitionId=None,
                actionSummary=f"Read '{sym}' -> Active states: {{{', '.join(lbls)}}}",
                status=("ACCEPTED" if has_acc else "REJECTED") if is_last else "RUNNING",
            )
        )

    accepted = any(s in machine.acceptStates for s in current_states)
    return SimulationResult(
        accepted=accepted,
        finalStatus="ACCEPTED" if accepted else "REJECTED",
        totalSteps=len(traces),
        traces=traces,
        message=f"Simulation finished. {'At least one branch reached accept state.' if accepted else 'No active branch reached an accept state.'}",
    )
